### PR TITLE
Add NVIDIA DALI GPU preprocessing guide

### LINKED ISSUE #20573: Title:** Request for Official NVIDIA DALI Pipeline for YOLO Preprocessing
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Hi Ultralytics team,  

Could you share an **official NVIDIA DALI pipeline** that exactly replicates YOLO’s preprocessing logic? Specifically:  
- **Letterbox resize** to 640x640 (preserve aspect ratio, add black padding `0`)  
- Normalization to `[0, 1]` range  
- RGB → CHW format conversion  

We’re integrating DALI for TensorRT deployment and need to ensure input consistency with YOLO’s training/inference pipeline. Could you provide a ready-to-use function like:  

```python
@dali.pipeline_def(batch_size=8, num_threads=4, device_id=0)
def pipe():
    # ... preprocessing steps ...
    return result
```  

This would save significant effort and avoid mismatches in input handling.  

Thanks!

### Additional

_No response_

### PR DESCRIPTION
**Pull request by raimbekovm**
Closes https://github.com/ultralytics/ultralytics/issues/20573

## Summary

- New guide: `docs/en/guides/nvidia-dali.md`
- DALI pipelines replicating YOLO LetterBox preprocessing (basic + centered)
- Integration with `model.predict()` via tensor passthrough
- Video stream processing with `fn.external_source`
- Triton Inference Server ensemble deployment (DALI + TensorRT)
- Updated `mkdocs.yml` navigation and `guides/index.md`

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
📚 This PR mainly expands and sharpens Ultralytics documentation, adding a new NVIDIA DALI guide for GPU-based YOLO preprocessing and updating Ultralytics Platform docs for clearer behavior around billing, API access, annotation tools, teams, and training views.

### 📊 Key Changes
- 🚀 Added a brand-new **NVIDIA DALI GPU Preprocessing** guide covering:
  - how to move YOLO preprocessing from CPU to GPU
  - matching Ultralytics `LetterBox` behavior with DALI
  - using DALI with `model.predict()`
  - video-stream preprocessing workflows
  - Triton Inference Server + TensorRT ensemble deployment
  - supported tasks, limitations, and FAQ
- 🧭 Added the new DALI guide to the docs navigation and guides index for easier discovery.
- 💳 Updated **Platform billing docs** to reflect the current cancellation flow:
  - cancel from **Settings > Plans**
  - added note about **Resume Subscription** before the billing period ends
  - clarified when Pro features and monthly credits stop
  - expanded transaction type descriptions
- 🕒 Clarified several **time-based platform behaviors**:
  - activity undo is available for **1 hour**
  - account export download links are valid for **1 hour**
  - team invites no longer expire after 7 days; they remain valid until accepted or canceled
- 👥 Improved **team management docs**:
  - pending invites count against seat limits
  - resending an invite rotates the token and sends a fresh email
- 🔐 Refined **Platform API docs** with clearer authentication rules:
  - emphasized which routes require an active **browser session**
  - clarified which routes do **not** support API-key auth
  - explained deployment route API-key support by endpoint
  - updated webhook docs to explain they are currently internal, not user-configurable
  - clarified cloning, prediction token, warmup, training, storage, and monitoring route behavior
- ✍️ Enhanced **annotation docs**:
  - documented **freehand polygon drawing** with `Shift + Move`
  - clarified polygon closing behavior
  - updated Smart mode interactions for SAM and mask add/subtract behavior
- 🖥️ Updated **Platform UI docs**:
  - training GPU availability now notes H200/B200 access for **Pro or Enterprise**
  - quickstart GPU recommendations were clarified
  - project comparison now documents **three view modes**: Cards, Compact, and Table
  - Smart mode shortcut now mentions **SAM or YOLO model**
- 🧹 Minor docs/navigation cleanup, including reordering guide entries in `mkdocs.yml`.

### 🎯 Purpose & Impact
- ⚡ Helps deployment-focused users speed up YOLO pipelines by showing how to remove CPU preprocessing bottlenecks with NVIDIA DALI.
- ✅ Makes advanced deployment setups easier to reproduce correctly, especially for **TensorRT** and **Triton** users.
- 🧠 Reduces confusion by aligning docs with actual Ultralytics Platform behavior for billing, invites, exports, and undo timing.
- 🔒 Improves developer experience by clearly explaining which API routes work with **API keys** versus browser-session authentication.
- 🎨 Makes annotation features easier to discover and use, especially for segmentation workflows and Smart mode editing.
- 📈 Gives Platform users clearer guidance on training hardware options and experiment comparison views.
- 🌍 Overall, this is a **documentation-focused PR** with low product risk but high usability value, helping both new and advanced users understand the platform and deploy YOLO models more effectively.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @raimbekovm, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This is an automated message to help with review readiness, and an engineer will assist you shortly. Please review the checklist below for a smooth integration 😊

-✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
-✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
-✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
-✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
-✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
-✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
-✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_  — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Good addition overall with useful coverage of DALI, Ultralytics integration, and Triton deployment, but there are a few correctness issues to fix before merging: two navigation indentation problems, a GPU→CPU→GPU copy in the video path, and the Triton section currently claims end-to-end detections even though the ensemble only returns raw model outputs.

💬 Posted 4 inline comments

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review 2

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The guide is well structured and covers a valuable deployment workflow, but there are a few correctness issues that need to be called out before merging. The main gap is that both the direct `model.predict()` example and the Triton ensemble path lose the original image geometry, so the returned boxes are still in letterboxed 640×640 coordinates. The batching advice for raw encoded images is also unsafe as written.

💬 Posted 3 inline comments

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review 3

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Useful and well-structured guide overall, but a few technical details should be tightened before merging: the reverse letterbox formula doesn't match Ultralytics' rounding behavior, the video example unnecessarily bakes in a CPU round-trip, and the TensorRT limitation is overstated.

💬 Posted 3 inline comments

File: docs/en/guides/nvidia-dali.md [hunk 1]
```diff
@@ -0,0 +1,608 @@
...
+        pipe.build()
+
+        cap = cv2.VideoCapture("video.mp4")
+        while cap.isOpened():
+            ret, frame = cap.read()
+            if not ret:
+                break
+
+            # Feed BGR frame (convert to RGB for DALI)
+            frame_rgb = cv2.cvtColor(frame, cv2.COLOR_BGR2RGB)
```
Comment (UltralyticsAssistant) [inline docs/en/guides/nvidia-dali.md:301]: 💡 **MEDIUM**: This video example reintroduces CPU work in the hot path with `cv2.cvtColor(...)`, and a few lines later it round-trips the batch through CPU again with `output.as_cpu().as_array()` before sending it back to CUDA. In a guide centered on eliminating CPU preprocessing bottlenecks, that is pretty misleading because it can erase much of the benefit for streaming use cases. I'd either move the color conversion into the DALI graph and use a GPU handoff mechanism, or clearly label this as a simple fallback example rather than a high-performance path. DALI's own examples and release notes emphasize direct PyTorch integration and zero-copy output paths instead. ([github.com](https://github.com/NVIDIA/DALI?utm_source=openai))

File: docs/en/guides/nvidia-dali.md [hunk 2]
```diff
@@ -0,0 +1,608 @@
...
+comments: true
+description: Learn how to use NVIDIA DALI for GPU-accelerated preprocessing with Ultralytics YOLO models. Eliminate CPU bottlenecks by running letterbox resize, padding, and normalization on the GPU for faster TensorRT and Triton deployments.
+keywords: NVIDIA DALI, GPU preprocessing, Ultralytics, YOLO, YOLO26, TensorRT, Triton Inference Server, letterbox, inference optimization, deep learning, computer vision, deployment, video processing, batch inference, DALI pipeline, CV-CUDA
+---
+
+# GPU-Accelerated Preprocessing with NVIDIA DALI
+
+## Introduction
+
+When deploying [Ultralytics YOLO](../models/index.md) models in production, [preprocessing](https://www.ultralytics.com/glossary/image-preprocessing) often becomes the bottleneck. While [TensorRT](../integrations/tensorrt.md) can run model [inference](../modes/predict.md) in under 1ms, the CPU-based preprocessing (resize, pad, normalize) can take 2-10ms per image, especially at high resolutions. [NVIDIA DALI](https://docs.nvidia.com/deeplearning/dali/user-guide/docs/index.html) (Data Loading Library) solves this by moving the entire preprocessing pipeline to the GPU.
```
Comment (UltralyticsAssistant) [inline docs/en/guides/nvidia-dali.md]: 📝 **LOW**: The hardcoded `under 1ms` claim overstates the current official YOLO26 TensorRT benchmark numbers in the repo README (for example, YOLO26n is listed at 1.7 ms on T4). That makes the intro read more like a best-case benchmark than a generally applicable rule of thumb. Please either qualify the exact model/hardware conditions or use a less absolute phrasing such as `low-single-digit milliseconds`. ([github.com](https://github.com/ultralytics/ultralytics?utm_source=openai))

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review 4

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Well-structured and useful guide overall, but I found two doc-accuracy issues worth fixing before merge: the intro currently overstates TensorRT latency, and the video-stream example keeps CPU work in the critical path despite positioning DALI as a zero-CPU preprocessing solution.

💬 Posted 2 inline comments

**Comment by raimbekovm** [comment]
@Y-T-G Ready for review when you have time.

File: docs/en/guides/nvidia-dali.md [hunk 3]
```diff
@@ -0,0 +1,608 @@
...
+
+    # Convert to numpy or PyTorch tensors
+    batch_np = output.as_cpu().as_array()  # Shape: (batch_size, 3, 640, 640)
+    print(f"Output shape: {batch_np.shape}, dtype: {batch_np.dtype}")
+    print(f"Value range: [{batch_np.min():.4f}, {batch_np.max():.4f}]")
+    ```
+
+## Using DALI with Ultralytics Predict
+
+You can pass a preprocessed [PyTorch](https://www.ultralytics.com/glossary/pytorch) tensor directly to `model.predict()`. When a `torch.Tensor` is passed, Ultralytics **skips image preprocessing** (letterbox, BGR→RGB, HWC→CHW, and /255 normalization) and only performs device transfer and [dtype](https://www.ultralytics.com/glossary/precision) casting before sending it to the model.
```
Comment (Y-T-G) [inline docs/en/guides/nvidia-dali.md]: The linked glossary for dtype is not related to dtype. Suggest removing it

File: docs/en/guides/nvidia-dali.md [hunk 4]
```diff
@@ -0,0 +1,608 @@
...
+
+    # Run inference through the ensemble
+    result = client.infer(model_name="ensemble_dali_yolo", inputs=[input_tensor])
+    detections = result.as_numpy("OUTPUT")  # Shape: (1, 300, 6) -> [x1, y1, x2, y2, conf, class_id]
+
+    # Filter by confidence (no NMS needed — YOLO26 is end-to-end)
+    detections = detections[0]  # First image
+    detections = detections[detections[:, 4] > 0.25]  # Confidence threshold
+    print(f"Detected {len(detections)} objects")
+    ```
```
Comment (Y-T-G) [inline docs/en/guides/nvidia-dali.md:523]: Ultralytics can load model deployed on Triton and handle this part automatically. Users simply pass images normally to `.predict()`

https://docs.ultralytics.com/guides/triton-inference-server/#running-inference

File: docs/en/guides/nvidia-dali.md [hunk 5]
```diff
@@ -0,0 +1,608 @@
...
+
+Yes. Use `DALIGenericIterator` to get preprocessed `torch.Tensor` outputs, then pass them to `model.predict()`. However, the performance benefit is greatest with [TensorRT](../integrations/tensorrt.md) models where inference is already very fast and CPU preprocessing becomes the bottleneck.
+
+### What is the difference between `fn.pad` and `fn.crop` for padding?
+
+`fn.pad` adds padding only to the **right and bottom** edges. `fn.crop` with `out_of_bounds_policy="pad"` centers the image and adds padding symmetrically on all sides, matching Ultralytics `LetterBox(center=True)` behavior.
+
+### Does DALI produce pixel-identical results to CPU preprocessing?
+
+Nearly identical. Set `antialias=False` in `fn.resize` to match OpenCV's `cv2.INTER_LINEAR`. Minor [floating-point](https://www.ultralytics.com/glossary/precision) differences (< 0.001) may occur due to GPU vs CPU arithmetic, but these have no measurable impact on detection [accuracy](https://www.ultralytics.com/glossary/accuracy).
```
Comment (Y-T-G) [inline docs/en/guides/nvidia-dali.md]: floating-point glossary link is not related. Suggest removing it

**Review by Y-T-G — CHANGES_REQUESTED** [review]
Nice writeup. Just need a few changes.

**Comment by raimbekovm** [comment]
@Y-T-G addressed all comments. for the triton part - added an info block explaining why tritonclient is used directly (DALI ensemble expects raw JPEG bytes while YOLO() sends preprocessed float32 tensors). could you verify this is correct?

**Review by Y-T-G — APPROVED** [review]

**Comment by glenn-jocher** [comment]
I pushed a small docs follow-up here to make the OpenCV `feed_input()` loop clearly a simple GPU->CPU->GPU fallback and to point readers to the reader/iterator path for real high-throughput GPU pipelines. Pending the current CI rerun, I do not see another technical blocker from my side.

**Comment by glenn-jocher** [comment]
I pushed one more docs refinement here in `afdb4b252`: the DALI pipeline section now keeps the centered `LetterBox(center=True)` version as the main reference and collapses the bottom-right `fn.pad(...)` variant into a short note. The goal was to tighten the guide without losing the distinction between exact Ultralytics parity and simpler custom pipelines.

**Comment by UltralyticsAssistant** [comment]
🎉 Huge thanks to @raimbekovm for this fantastic merge, with valuable contributions from @fcakyon, @Y-T-G, @Laughing-q, @onuralpszr, and @glenn-jocher!

> “The only source of knowledge is experience.” — Albert Einstein

This PR is a great example of that idea in action: turning real platform and deployment experience into clearer, more actionable docs for the whole community. From the new NVIDIA DALI guide to the many important Ultralytics Platform clarifications around billing, API behavior, annotation workflows, teams, and training views, this work makes YOLO deployment and day-to-day usage more intuitive for everyone.

Really appreciate the care, precision, and user-first thinking behind this one — a meaningful docs improvement with broad impact. 🙌
